spec.emu

```diff
@@ -168,3 +168,334 @@ contributors: Surma, Dan Ehrenberg
     <dd>Initializes a Module Block with host-dependent information.</dd>
   </dl>
 </emu-clause>
+
+
+<emu-clause id="sec-executable-code-and-execution-contexts">
+  <h1>Executable Code and Execution Contexts</h1>
+
+  <emu-clause id="sec-code-realms">
+    <h1>Realms</h1>
+    <p>Before it is evaluated, all ECMAScript code must be associated with a <dfn id="realm" variants="realms">realm</dfn>. Conceptually, a realm consists of a set of intrinsic objects, an ECMAScript global environment, all of the ECMAScript code that is loaded within the scope of that global environment, and other associated state and resources.</p>
+    <p>A realm is represented in this specification as a <dfn id="realm-record" variants="Realm Records">Realm Record</dfn> with the fields specified in <emu-xref href="#table-realm-record-fields"></emu-xref>:</p>
+    <emu-table id="table-realm-record-fields" caption="Realm Record Fields" oldids="table-21">
+      <table>
+        <tr>
+          <th>
+            Field Name
+          </th>
+          <th>
+            Value
+          </th>
+          <th>
+            Meaning
+          </th>
+        </tr>
+        <tr>
+          <td>
+            [[Intrinsics]]
+          </td>
+          <td></td>
+          <td></td>
+        </tr>
+        <tr>
+          <td>
+            [[GlobalObject]]
+          </td>
+          <td></td>
+          <td></td>
+        </tr>
+        <tr>
+          <td>
+            [[GlobalEnv]]
+          </td>
+          <td></td>
+          <td></td>
+        </tr>
+        <tr>
+          <td>
+            [[TemplateMap]]
+          </td>
+          <td></td>
+          <td></td>
+        </tr>
+        <tr>
+          <td>
+            <ins>[[ModuleMap]]</ins>
+          </td>
+          <td>
+            <ins>a List of Record { [[ReferencingScriptOrModule]]: Script Record or Module Record or *null*, [[Specifier]]: String or Object, [[ModuleRecord]]: Module Record }</ins>
+          </td>
+          <td>
+            <p><ins>Module Records are cached separately per Realm: two imports from the same realm using the same specifier from the same referencing script or module must resolve to the same Module Record. The cache needs to be per-Realm to support cases when [[ReferencingScriptOrModule]] is *null*.</ins></p>
+            <p><ins>This List respects the following invariants:</ins></p>
+            <ul>
+              <li>
+                <ins>It does not contain two Records _record1_ and _record2_ such that _record1_.[[ReferencingScriptOrModule]] is _record2_.[[ReferencingScriptOrModule]] and _record1_.[[Specifier]] is _record2_.[[Specifier]].</ins>
+              </li>
+              <li>
+                <p><ins>For every Record _record_ it contains, if Type(_record_.[[Specifier]]) is Object then:</ins></p>
+                <ul>
+                  <li>
+                    <ins>_record_.[[Specifier]] has a [[ModuleBlockBody]] internal slot.</ins>
```

Comment:
This is one way to do it, but another way might be to actually just go ahead and split up the module map into the two separate maps for compilation and instances. Eg if all resolutions are based on compiled module returns, then the import of a module block is not the cached resolution of a module block specifier, but rather just the instance lookup for a compiled module. Would something like that make sense? Not tied to either way just trying to work through where this goes.

Reply:
Potentially yes, but I first want to figure out exactly how caching should work. For example, currently the proposal does not guarantee that the two imports in this file evaluate to the same module:
```js
import * as ns from "./file.js";

await import(module {
  import * as ns from "./file.js";
});
```

Reply:
See https://github.com/tc39/proposal-js-module-blocks/pull/68#discussion_r912411474 for discussion on this.